### PR TITLE
Fixes terminal completion command for odo

### LINKED ISSUE #5536: Errors while sourcing output of `odo utils terminal <shell>`
/kind bug

## What versions of software are you using?

**Operating System:**
```
❯ cat /etc/fedora-release
Fedora release 35 (Thirty Five)
```

**Output of `odo version`:**
`odo v2.5.0` (9e09d80e5)

## How did you run odo exactly?
```shell
source <(odo utils terminal zsh)
```

Also tried to source the corresponding output after switching my shell to Bash:
```shell
source <(odo utils terminal bash)
```

## Actual behavior
The `odo utils` command is unable to set the shell prompt to the correct application and component names.
The prompt is set to `<no application>/<no component>` while it should not.
Below is an excerpt of the logs. Full logs in the last section below.

```
asoro@work-laptop:~/w/t/multi-module-devfile$ source <(odo utils terminal zsh)                                                                            
--- OMITTED FOR BREVITY ---                                                                                                                                                                                                                                                                          
 ✗  unknown shorthand flag: 'q' in -q
--- OMITTED FOR BREVITY ---          
 ✗  unknown flag: --skip-connection-check
[<no application>/<no component>]asoro@work-laptop:~/w/t/multi-module-devfile$ 
```

There are 2 issues reported here:
- ` ✗  unknown shorthand flag: 'q' in -q`
- `✗  unknown flag: --skip-connection-check`

## Expected behavior

I am not sure if we will keep the `odo utils` subcommand in v3.
If so, it looks like the commands used to determine both the [application](https://github.com/redhat-developer/odo/blob/main/pkg/odo/cli/utils/terminal.go#L25) and [component](https://github.com/redhat-developer/odo/blob/main/pkg/odo/cli/utils/terminal.go#L32) need to be refreshed / tested:

```shell
APP=$(odo application get -q --skip-connection-check)
COMPONENT=$(odo component get -q --skip-connection-check)
```

## Any logs, error output, etc?

Full logs below:

```
asoro@work-laptop:~/w/t/multi-module-devfile$ source <(odo utils terminal zsh)                                                                                                                                                      
Error: unknown shorthand flag: 'q' in -q                                                                                                                                                                                                       
Usage:                                                                                                                                                                                                                                         
  odo [flags]                                                                                                                                                                                                                                  
  odo [command]                                                                                                                                                                                                                                
                                                                                                                                                                                                                                               
Examples:                                                                                                                                                                                                                                      
  # Creating and deploying a Node.js project                                                                                                                                                                                                   
  git clone https://github.com/odo-devfiles/nodejs-ex && cd nodejs-ex                                                                                                                                                                          
  odo create nodejs                                                                                                                                                                                                                            
  odo push                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                               
  # Accessing your Node.js component                                                                                                                                                                                                           
  odo url create                                                                                                                                                                                                                               
                                                                                                                                                                                                                                               
Commands:                                                                                                                                                                                                                                      
  catalog      Catalog related operations (describe, list, search)                                                                                                                                                                             
  component    Manage components (create, list, push, v2delete)                                                                                                                                                                                
  preference   Modifies preference settings (registry, set, unset, view)                                                                                                                                                                       
  project      Perform project operations (create, delete, get, list, set)
  url          Expose component to the outside world (create, delete, list)

Utility Commands:
  build-images Build images
  deploy       Deploy components
  dev          Deploy component to development cluster
  help         Help about any command
  login        Login to cluster
  logout       Log out of the current OpenShift session
  utils        Utilities for terminal commands and modifying odo configurations (terminal)
  version      Print the client version information

Component Shortcuts:
  create       Create a new component 
  list         List all components in the current application 
  push         Push source code to a component 
  v2delete     Delete component 

Flags:
      --complete             Install completion for odo command
  -h, --help                 Help for odo
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
      --uncomplete           Uninstall completion for odo command
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
  -y, --y                    Don't prompt user for typing 'yes'

Use "odo [command] --help" for more information about a command.

 ✗  unknown shorthand flag: 'q' in -q
Error: unknown flag: --skip-connection-check
Usage:
  odo component get [flags]

Examples:
  # Get the currently active component
  odo component get

Flags:
      --context string   Use given context directory as a source for component settings
  -h, --help             Help for get
      --project string   Project, defaults to active project
  -q, --short            If true, display only the component name

Additional Flags:
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging

 ✗  unknown flag: --skip-connection-check
[<no application>/<no component>]asoro@work-laptop:~/w/t/multi-module-devfile$ 
```

### PR DESCRIPTION
**Pull request by cdrage**
**What type of PR is this:**

/kind cleanup

**What does this PR do / why we need it:**

This PR:
* Changes the command `odo utils terminal` to `odo completion` similar
  to kubectl.
* Re-enables / fixes the terminal completion command support for bash,
  zsh, fish and powershelll

**Which issue(s) this PR fixes:**

Fixes #5536

**How to test changes / Special notes to the reviewer:**

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

**Comment by cdrage** [comment]
If people can please test please do!

The instructions are at `odo completion --help` for bash, zsh, powershell and fish.

**Comment by valaparthvi** [comment]
@cdrage, do you think you can add a documentation for this command?

File: pkg/odo/cli/completion/completion.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	## Load persistently
+	%[1]s completion fish > ~/.config/fish/completions/%[1]s.fish
+
+	# POWERSHELL
+
+	## Load into your current shell environment
+	%[1]s completion powershell | Out-String | Invoke-Expression
+
+	## Load persistently
+	%[1]s completion powershell >> $PROFILE
```
Comment (valaparthvi) [inline pkg/odo/cli/completion/completion.go:53]:
```suggestion
	%[1]s powershell | Out-String | Invoke-Expression

	## Load persistently
	%[1]s powershell >> $PROFILE
```

**Comment by rm3l** [comment]
@cdrage I reproduced what I was mentioning during the Standup call. Unlike the other additional flags, `-o` does not seem to be listed in the completion list. Don't know if that is the expected behavior with this flag.
I am using zsh, FYI.

Example with `odo analyze`:
```
❯ odo analyze -<TAB>
--kubeconfig      -- Paths to a kubeconfig. Only required if out-of-cluster.                                                                                                                                                                                                   
--v           -v  -- number for the log level verbosity. Level varies from 0 to 9 (default 0).                                                                                                                                                                                 
--vmodule         -- comma-separated list of pattern=N settings for file-filtered logging
```

And here is the help output of `odo analyze`, which mentions the `-o` option.:
```
❯ odo analyze --help 
Detect devfile to use based on files present in current directory

Usage:
  odo analyze [flags]

Flags:
  -h, --help   Help for analyze

Additional Flags:
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
  -o, --o string             Specify output format, supported format: json
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
```

Same behavior with `odo describe component` or `odo describe binding`.

**Review by valaparthvi — CHANGES_REQUESTED** [review]

**Comment by cdrage** [comment]
> @cdrage I reproduced what I was mentioning during the Standup call. Unlike the other additional flags, `-o` does not seem to be listed in the completion list. Don't know if that is the expected behavior with this flag. I am using zsh, FYI.
> 
> Example with `odo analyze`:
> 
> ```
> ❯ odo analyze -<TAB>
> --kubeconfig      -- Paths to a kubeconfig. Only required if out-of-cluster.                                                                                                                                                                                                   
> --v           -v  -- number for the log level verbosity. Level varies from 0 to 9 (default 0).                                                                                                                                                                                 
> --vmodule         -- comma-separated list of pattern=N settings for file-filtered logging
> ```
> 
> And here is the help output of `odo analyze`, which mentions the `-o` option.:
> 
> ```
> ❯ odo analyze --help 
> Detect devfile to use based on files present in current directory
> 
> Usage:
>   odo analyze [flags]
> 
> Flags:
>   -h, --help   Help for analyze
> 
> Additional Flags:
>       --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
>   -o, --o string             Specify output format, supported format: json
>   -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
>       --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
> ```
> 
> Same behavior with `odo describe component` or `odo describe binding`.

Ah! That's a bug, I'll find it. Thank you. Odd that it's only `-o` doing it.

**Review by valaparthvi — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/cli.go [hunk 2]
```diff
@@ -152,10 +152,6 @@ func odoRootCmd(name, fullName string) *cobra.Command {
 	_ = pflag.CommandLine.MarkHidden("skip_headers")
 	_ = pflag.CommandLine.MarkHidden("skip_log_headers")
 
-	// We will mark the command as hidden and then re-enable if the command
-	// supports json output
-	_ = pflag.CommandLine.MarkHidden("o")
```
Comment (feloy) [inline pkg/odo/cli/cli.go:157]: By removing this, the option now appears in all commands, event if they do not support it.

**Comment by cdrage** [comment]
@feloy @rm3l 

Fixed!

Basically... I've unhidden the -o flag, but I've modified the help usage page to hide it if it's not supported.

Essentially, we cannot "unhide" and "hide" flags programatically within Cobra in order to generate the correct bash completion.

Unfortunately, with the way that completion is handled within cobra, it'll still show for each command.

```sh
go/src/odo  fix-utils ✗                                                                                                                                                                                                                                                                                                                                         3d21h ⚑
▶ odo deploy --o json
 ✗  Machine readable output is not yet implemented for this command
```

Ex. For when trying to run it.

I suggest that we still merge this in, and create an issue upstream (Cobra) as well as an issue within our repo with a WONTFIX until it's fixed upstream.

**Comment by feloy** [comment]
You will need to fix these "validate" errors:

```
pkg/odo/cli/completion/completion.go:72:33: Error return value of `(*github.com/spf13/cobra.Command).GenBashCompletion` is not checked (errcheck)
				cmd.Root().GenBashCompletion(os.Stdout)
				                            ^
pkg/odo/cli/completion/completion.go:79:14: Error return value of `out.Write` is not checked (errcheck)
				out.Write([]byte(zsh))
				         ^
pkg/odo/cli/completion/completion.go:80:32: Error return value of `(*github.com/spf13/cobra.Command).GenZshCompletion` is not checked (errcheck)
				cmd.Root().GenZshCompletion(out)
				                           ^
pkg/odo/cli/completion/completion.go:82:33: Error return value of `(*github.com/spf13/cobra.Command).GenFishCompletion` is not checked (errcheck)
				cmd.Root().GenFishCompletion(os.Stdout, true)
				                            ^
pkg/odo/cli/completion/completion.go:84:47: Error return value of `(*github.com/spf13/cobra.Command).GenPowerShellCompletionWithDesc` is not checked (errcheck)
				cmd.Root().GenPowerShellCompletionWithDesc(os.Stdout)
```

**Review by feloy — APPROVED** [review]

**Comment by cdrage** [comment]
Ready for review 👍 @rm3l @valaparthvi

**Review by rm3l — APPROVED** [review]
Tested with both ZSH and Bash, and completion works great 🎉

One side note: we could have added this to the documentation (command reference and/or somewhere in the installation guide). For now, this seems hidden as a feature.

/approve

**Comment by cdrage** [comment]
> 

You're right, I'll add a doc for that!

**Pull request merged**